code/__HELPERS/atoms.dm

```diff
@@ -323,3 +323,17 @@ rough example of the "cone" made by the 3 dirs checked
 	if(!storage_datum)
 		return
 	. += storage_datum.real_location?.resolve()
+
+/// Returns an x and y value require to reverse the transformations made to center an oversized icon
+/atom/proc/get_oversized_icon_offsets()
+	if (pixel_x == 0 && pixel_y == 0)
+		return list("x" = 0, "y" = 0)
+	if (!(type in GLOB.oversized_atom_icon_offsets))
+		var/icon/my_icon = icon(icon, icon_state)
+		var/icon_width = my_icon.Width()
+		var/icon_height = my_icon.Height()
+		GLOB.oversized_atom_icon_offsets[type] = list(
+			"x" = icon_width > world.icon_size && pixel_x != 0 ? (icon_width - world.icon_size) * 0.5 : 0,
+			"y" = icon_height > world.icon_size && pixel_y != 0 ? (icon_height - world.icon_size) * 0.5 : 0,
+		)
+	return GLOB.oversized_atom_icon_offsets[type]
```

Comment:
icon seems like a safer bet, since that could change mid go

Reply:
There's no perfect solution but typepath seemed the most sensible key to use.
Not everything which shares the same icon file necessarily shares the same offset (detected by whether it has set a pixel_x/pixel_y).
If I combine the icon file and icon state into a key I end up with way more list entries than necessary.
Anything with the same typepath probably at least always is going to want a similar offset regardless of what icon it is using.

Reply:
just cache the width and height instead of caching the full offset
that calculation is cheap the expensive part is the Width() and Height() calls

Reply:
Oh yeah that makes a lot of sense, I will go back and use the same cache in some other places too while I am at it.